docs: Add README.md (#11)
perun-network/erdstall-ts-sdk: README.md

> # Erdstall typescript SDK

Reviewer: ```suggestion
# Erdstall TypeScript SDK
```

**What is Erdstall?**&emsp;<!--

> -->[Erdstall](https://erdstall.dev) is a layer-2 platform for Ethereum developed by [PolyCrypt GmbH](https://polycry.pt) based on research by the chair of applied cryptography at the technical university of Darmstadt, Germany.

Reviewer: ```suggestion
-->[Erdstall](https://erdstall.dev) is a layer-2 platform for Ethereum developed by [PolyCrypt](https://polycry.pt) based on research by the chair of applied cryptography at the technical university of Darmstadt, Germany.
```

It uses a [TPM](https://en.wikipedia.org/wiki/Trusted_Platform_Module) — such as Intel SGX or ARM TrustZone — ensuring reliable and trustworthy execution of the platform without the need for public verification of all computations (in comparison to accumulated zero-knowledge proofs, TPMs allow the generation of more compact and less computationally expensive proofs, thereby scaling much better and consuming less power).

> In Erdstall, the TEE operator periodically generates checkpoints, also called balance proofs, that can be used to withdraw funds on-chain.

Reviewer: ```suggestion
In Erdstall, the TEE operator periodically generates checkpoints, also called balance proofs, that can be used to exit the system if the operator were to go offline.
```

Meanwhile, market operators benefit from fast throughput and low operation costs, innate user-trust, and fast and cheap deposits and withdrawals via Erdstall.
Another benefit is that content creators can natively mint NFTs off-chain, and the cost of on-chain token creation only happens when the NFT is withdrawn by a user into the Ethereum mainnet.
Because our protocol is very light-weight, it is easy to integrate into existing solutions.

**Scope of this SDK**&emsp;<!--
-->This SDK is designed to offer the complete client-side functionality of Erdstall.
It natively integrates with any ethers.js provider, such as Metamask, as its wallet manager.
The SDK offers full control, but also packages the most important functionality into an easy-to-use API.
The SDK is designed such that code snippets that use Erdstall can be written generically to be reused within any site using Erdstall.
That way, common web components, such as payment/donation buttons, can easily be reused on your own homepage.

## Getting started

To get your Erdstall app started, there are a few steps you need to take.

> **Importing Erdstall**&emsp;<!--

Reviewer: Add sentence here (will open PR for this soon)
> The Erdstall SDK is not published as a node package yet, so clone it with `git clone https://github.com/perun-network/erdstall-ts-sdk.git erdstall`, then `cd erdstall; yarn link` to locally register the erdstall node module.

-->First, import the SDK's main source directory:

```ts

> import * as SDK from "erdstall-ts-sdk/src"

Reviewer: ```suggestion
import * as erdstall from 'erdstall'
```

Please name the package variable `erdstall` and the client instance `client` below. This is more in line with how you'd usually call this. `SDK` is non-descriptive.

Reviewer: I'm about to open a PR that makes this a node package, after which we can use it this way.

```

**Creating the client object**&emsp;<!--

> -->The SDK is straight-forward to use: everything you need is neatly bundled in the `SDK.Erdstall` interface, which you can instantiate via `SDK.NewClient()`, as follows:

Reviewer: rename `SDK` to `erdstall` everywhere

```ts
// Initialize these fields with values received from your ethers or web3 setup.

> var ethAccount: {address: SDK.ledger.Address, signer: ethers.Signer};

Reviewer: Do we have a TS type for that?

Author: no. This was just to semantically group those values together while also writing down their types.

// Create client object.

> let Erdstall = SDK.NewClient(

Reviewer: ```suggestion
let client = SDK.NewClient(
```
or `erdclient`.

Reviewer: In particular, local variables of class instances usually don't start with a capital letter.

	ethAccount.address,
	ethAccount.signer,
	new URL("https://operator.erdstall.dev"));
// Set up any event listeners needed by the application.

> Erdstall.once("open", () => { myDApp.run(Erdstall); });

Reviewer 2: Try to avoid using a string when you can use a constant instead. Otherwise it is easy to introduce bugs when maintaining or changing the code.

You can replace `"open"` by a constant `Erdstall.OpenEvent`.

Even better would be to have a specific handler for each event type: `Erdstall.OnOpen`.

Author: Typescript, as far as I know, has a compile time check to check the validity of the string.

Reviewer: See below comment, I generally agree, bu this is just the README PR.

// Establish the client's connection.
Erdstall.initialize();
```

**Setting up events**&emsp;<!--
-->After creation, the Erdstall client is now only partially initialized.
Now is the time to set up any `SDK.ledger.ErdstallEvent` and `SDK.enclave.EnclaveEvent` handlers using `Erdstall.on()` and `Erdstall.once()`.
The following `EnclaveEvent` events exist (emitted event types are written in parentheses):

> * **`"open"`**:

Reviewer 2: See comment above. You should at least use constants for the event names. Even better would be to have a separate handler function for each event type. See [jQuery Events](https://api.jquery.com/category/events/), for example.

Reviewer 3: Typescript has type narrowing, which means strings like that are evaluated at compile time. It is not possible to use this API with non-existent events or similar.

Personally I dislike the usage of these `string`s, but it is common in every other web3-provider/library, so I did not want to break with the flow and keep it as ordinary as possible.

I also plan to use some variation of dependent types such that users of this API, given that they use some form of auto-completion, are given the correct eventhandler signature as soon as an event is specified. But I wanted to do this in the second iteration.

Reviewer: These are issues that concern the sdk in general and not this README PR. I agree that string constants are nicer, and specific methods are best. But let's just get a first version of the README up and move these discussions into separate issues.

* `"exitproof"(SDK.BalanceProof)`:
		triggered when a balance proof intended for exiting Erdstall on-chain is received from a proof subscription.
	This happens after a call to `Erdstall.exit()` or during `Erdstall.leave()`.

The following `ErdstallEvent` smart contract events exist:

* **`"Deposited"`**:
		you deposited money into Erdstall.
* **`"Withdrawn"`**:
		you withdrew funds from Erdstall.
* `"Frozen"`:
		an unrecoverable error occurred and the Erdstall system is now frozen forever, and funds must be withdrawn using the latest "sealed" balance proof.
* `"Challenged"`:
		someone challenged the Erdstall operator to prove it is still alive.
* `"ChallengeResponded"`:
		the Erdstall operator successfully proved it still operates.
* `"TokenTypeRegistered"`:
		a new token type (such as ETH, ERC20, ERC721, etc.) has been registered with Erdstall on-chain.
* `"TokenRegistered"`:
		a specific token contract has been registered on-chain.

**Finishing up**&emsp;<!--
-->After all events have been set up, call `Erdstall.initialize()` to establish the client's connection to the Erdstall ledger.
You're now good to go!
The `open` event will be triggered and your logic will launch.

> ## Using the Erdstall client

Reviewer 2: It would be great to add a reference to a minimal working example.

Reviewer: created new issue #13

**Off-chain trading**&emsp;<!--
-->Erdstall natively supports digital asset trades in the form of fungible and non-fungible tokens (FT, NFT).
All trading requires an external marketplace that builds on-top of Erdstall.
The marketplace's role is to facilitate order match-making and to track the currently open offers.
Trading works as follows:

1. The seller creates a trade offer — which is a promise indicating he is willing to trade certain assets for a certain price — and sends it to a market operator.
2. The market operator broadcasts the trade offer to potential buyers, who can respond with a trade agreement, indicating they are willing to match the offer.
3. The market operator sends both the trade offer and the trade agreement to Erdstall.

>*This is currently not implemented, we still need to create a more unified API on how to connect the Erdstall client to any Market operator, both for proposing and accepting trades.*

**Off-chain NFT minting**&emsp;<!--
-->`Erdstall.mint(SDK.ledger.Address, SDK.Uint256) TXReceipt` can be used to create a new token of the specified kind and with the specified ID.

> The NFT can then be traded freely, and is only minted on-chain upon withdrawal.

Reviewer: Please add copyright and license notice to the bottom like in our other repos.
